docs/kb/wasm.rst

```diff
@@ -0,0 +1,307 @@
+
+.. code-block:: bash
+
+    asc fib.ts --textFile fib.wat --optimize
+
+The compiled example can be viewed at test/cql-pytest/test_wasm.py::test_docs_assemblyscript
+
+Similarly to C, the AssemblyScript can only be conveniently used with ``RETURNS NULL ON NULL INPUT`` UDFs that only have Wasm-compatible arguments/returns.
+
+Generating WAT from Wasm
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+For those who want to use precompiled Wasm modules, it's enough to translate Wasm bytecode to ``wat`` representation. On Linux, it can be achieved by a ``wasm2wat`` tool, available in most distributions in the ``wabt`` package.
+
+Example
+-------
+
+Here's how a ``wasm`` function can be declared:
+
+.. code-block:: cql
+
+    CREATE FUNCTION ks.fib (input bigint) RETURNS NULL ON NULL INPUT RETURNS bigint LANGUAGE wasm AS '
+    (module
+        (func $fib (param $n i64) (result i64)
+            (if
+            (i64.lt_s (local.get $n) (i64.const 2))
+            (return (local.get $n))
+            )
+            (i64.add
+            (call $fib (i64.sub (local.get $n) (i64.const 1)))
+            (call $fib (i64.sub (local.get $n) (i64.const 2)))
+            )
+        )
+        (memory (;0;) 2)
+        (export "memory" (memory 0))
+        (export "fib" (func $fib))
+        (global (;0;) i32 (i32.const 1024))
+        (export "_scylla_abi" (global 0))
+        (data $.rodata (i32.const 1024) "\01")
+    )'
+
+and it can be invoked just like a regular UDF:
+
+.. code-block:: cql
+
+    scylla@cqlsh:ks> CREATE TABLE t(id int, n bigint, PRIMARY KEY(id,n));
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 0);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 1);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 2);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 3);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 4);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 5);
+    scylla@cqlsh:ks> SELECT n, ks.fib(n) FROM t;
+
+    n  | ks.fib(n)
+    ----+-----------
+     0 |         0
+     1 |         1
+     2 |         1
+     3 |         2
+     4 |         3
+     5 |         5
+
+    (11 rows)
+
+Experimental status
+-------------------
+
+WebAssembly UDFs are still experimental due to insufficient testing. If backwards incompatible changes
+to the ABI are implemented in the future, they should be submitted as new ABI-versions, and use the same
+``LANGUAGE wasm`` clause in the CQL statements.
+
+ABI versions
+------------
+
+Different programming languages may require different ABIs. To support that, the Wasm program is required to
+export the symbol ``_scylla_abi``, that is a WebAssembly global with a 32-bit value of the offset in memory,
+where the version number can be read (that's the only method of exporting a constant in Rust).
+
+Currently, the only available ABI versions are 1 and 2. Both of them use the same protocol for passing
+parameters and returning values, but they differ in approaches to memory management.
```

Comment:
Reading the above sections which used ABI 1 instead of 2, I got the impression that version 1 and 2 also differ in the need to "serialize" and/or "byte-swap" values, and also the handling of nulls. Was this impression incorrect?

Reply:
All the things that you've mentioned are actually the same in both ABIs. I see now that I made the examples a bit confusing, because the example with "byte-swaps" is for a `CALLED ON NUL INPUT` UDF, and the simple example is for a `RETURNS NULL ON NULL INPUT` UDF.

If the first example wasn't `CALLED ON NUL INPUT`, it would not require the "byte-swaps", but it would still need to include `_scylla_malloc` and `_scylla_free` implementations. 

I'll try to highlight that difference in the rebase, so that users don't get the same impression, but I'm not sure if it will be enough to stop being confusing